### DISCUSSION TITLE
Case sensitive macro names?

### DISCUSSION THREAD

**Discussion by hamishwillee** [mdn/mdn-community#448] [Content]
Yari is case insensitive with respect to macro names - so in the docs we have  `{{glossary("Xxxxx")}}` , `{{Glossary("Xxxxx")}}`, and {{Deprecated_Header}} and {{deprecated_header}}, and so on.

Josh has started looking at consistency for the naming in JavaScript section: https://github.com/mdn/content/pull/29008 and https://github.com/mdn/content/pull/29006. The selected format is the most common used, which _probably_ matches https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Macros/Commonly_used_macros.

Consistency makes me happy, but since the macros are MDN-wide I think it is worth making sure we are all happy with both making this change, and the actual agreed format.

My preference is:
- We do only allow one case-sensitive format
- That format is all lower case (snake case). That is the easiest format to remember as an author, and it makes no difference to users.
- We plan to disallow other formats and make them a flaw.

**Comment by Josh-Cena** [comment]
I prefer lowercase for xref macros (`jsxref`, `domxref`, `cssxref`) and snake case (`optional_inline`, `deprecated_header`), but I prefer the "real case" for all CamelCase macros, like `HTMLElement`, `HTTPHeader`, `PreviousNext`, `SeeCompatTable`, etc. because those are much easier to read :( Furthermore, @onkarruikar would need to change the BCD sync script if we agree to use `deprecated_header` instead of `Deprecated_Header`.

FYI, here's the macro count on main (for the JS section):

<img width="541" alt="image" src="https://github.com/mdn/mdn-community/assets/55398995/559d7969-6805-4a5a-8449-98e921e3c8bd">

And here's the set of macro casings I settled on tentatively:

<img width="541" alt="image" src="https://github.com/mdn/mdn-community/assets/55398995/275b0627-92cb-4ef2-a334-97c8897409f3">

**Reply by hamishwillee** [reply]
FWIW Its a real pity we didn't settle on sake case or camel case when this was started. The mix is what makes it hard. The best option would be to rename anything that needs to be to be snake case. But that would be disruptive. If we're going to to do this I'd still prefer one rule - so: htmlelement, httpheader, previousnext, seecompattable.

Doing it at all is disruptive - once you decide that it is worth doing you choose the right way and suck up any pain :-).

**Reply by Josh-Cena** [reply]
I would prefer doing it like how we actually case things in a programming language: you use either snake_case or CamelCase, and it's fine if there's a mishmash of both (for legacy reasons), but you don't do `seecompattable`. Again, I would be overjoyed if we agree to change `seecompattable` to `see_compat_table`.

**Comment by Josh-Cena** [comment]
I would be very happy if (a) all macros can use snake_case (b) they are all valid JS identifiers, but that sounds far too disruptive.

**Comment by OnkarRuikar** [comment]
I would love to see `CamelCase` because it's more compact than `snake_case` and it stands out, e.g.
```
{{embed_live_sample("example_showing_some_thing", "100%", "300px"}}
vs
{{EmbedLiveSample("example_showing_some_thing", "100%", "300px"}}
```

I did try to make macro calls consistent before, tried changing to camel case. But changing massive amount of files just to change case was not welcomed.\
We have grown accustomed to using different cases for each macro, e.g. `jsxref`, `HTTPHeader`, and `deprecated_header`. So changing to fixed case will cause some discomfort. In the beginning we could allow people to use whatever case they want and update the macro calls at the end of the day in our nightly automated linting. Eventually when people get used to the case then we can start enforcing the case in Husky and PR workflows.

> We plan to disallow other formats and make them a flaw.

Using [our custom linter](https://github.com/mdn/content/blob/124d5c690925e723854558c6dd774728135a2c00/.markdownlint-cli2.jsonc#L126-L218) will be more effective than the Yari flaws mechanism.

***

As we are on the topic of prettying the macro calls, we need to iron out following issues as well:
- `{{ CSSref }}` vs `{{CSSref}}`: Someone was/is adding spaces around the macro names to make them more readable. But this creates conflict with the automation scripts because the usage is not fixed.
- `{{CSSref()}}` vs `{{CSSref}}`: We can get rid of `()` in cases where there are no arguments.
- `{{Macro("arg")}}` vs `{{Macro('arg')}}`: Can we choose to have only double quotes?
- Use of `px` or `%` in live sample macro calls and other inconsistencies: 
```
{{EmbedLiveSample('audio', '100%', '70px')}}
{{EmbedLiveSample('audio', '100%', '70')}}

{{EmbedLiveSample('audio', 250, 250)}}
{{EmbedLiveSample('audio', '250', '250')}}

{{EmbedLiveSample('audio', '', '70px')}}
{{EmbedLiveSample('audio', '100%', '70px')}}

{{EmbedLiveSample('audio_example')}}
{{EmbedLiveSample('audio example')}}
```

**Reply by Josh-Cena** [reply]
@OnkarRuikar These are all issues fixable by my checker, which does a full parsing of the macro content and runs Prettier over the expression. I plan to roll out these changes one by one.

**Comment by bsmth** [comment]
Thanks for kicking off the discussion, it's frustrating to grep through for matches as we have so many different formats. No preference on the casing, but it would be great to stick to **one format per macro**. The same sentiment on the list Onkar has pointed out; it would be great to be inflexible on these cases.

__No preference:__
I'm happy with what others decide for these:

* lower / mixed / CamelCase
* singe v double quotes
* `%` v `px`


__Preference:__
I think the following makes sense:

* disallow brackets when no arguments
* whitespace (+1 to get rid of it when not needed)
  * we have multiline macro calls, this would be good to make single-line

**Reply by yin1999** [reply]
Just to draw attention: we also have different formats for the `width` and `height` parameters of the livesample macro (quoted/unquoted numbers).

- `{{EmbedLiveSample("some_id", 120, 150)}}`
- `{{EmbedLiveSample("some_id", "120", "150")}}`

**Comment by wbamberg** [comment]
For naming, I prefer snake case to camel case, but I don't really care as long as we don't have more than one naming convention. I don't want to have to remember which convention to use for which macro, that's just ridiculous. If we're going to do that we might as well leave things in the current mess and save a lot of grief.

>> The best option would be to rename anything that needs to be to be snake case.

Yeah, I agree, this is best, and next best, if we don't want to rename, would be just to adopt snake case.

>> But that would be disruptive.

For renaming: the disruptiveness I think is mostly in two parts:
1. making and reviewing additional diffs, to change the places that renamed macros are called
2. getting people used to using new names

For the first of these, I think the additional disruption on top of just snake-casing is not that great, just because most of our invocations use at least one uppercase letter. I made a spreadsheet: https://docs.google.com/spreadsheets/d/1oXrwRhLRM745z0QOAm53ezaPcI55SId9gsFV4ffA7gQ/edit?usp=sharing that lists all the different macro invocations in all our docs, and tries to assess for each one:
- whether it would need to be changed to implement snake case
- whether it would need to be changed to implement renaming only (meaning, if we didn't need to change it anyway for snake-casing)

You can see most of the diffs - 71370 - are incurred in snake-casing, and only an additional 3105 in renaming.

(This also shows that just snake-casing is a BIG change, and hard for humans to review reliably. IMO we should use machines to test the change instead of relying on human review.)

One open question there is whether we rename xrefs (e.g. domxref to dom_xref) - @Josh-Cena above proposed we make an exception for xrefs. For these numbers I excluded xrefs, but that would add a lot more churn.

The second part of disruptiveness - getting people used to using new names - is harder to quantify, but I think it would be OK. We will get used to the names quickly enough.

But renaming also tends to open another worm-can. Why are some of our xrefs called `thingxref` and some aren't? Why are some of our sidebars called `thingsidebar` but some aren't? Why are some of our banner/header macros called `thingheader` and some aren't?

**Reply by Josh-Cena** [reply]
> One open question there is whether we rename (e.g. domxref to dom_xref)

Yes, I would exclude xrefs. They are sufficiently terse and adding an underscore doesn't seem worth it.

> Why are some of our sidebars called thingsidebar but some aren't?

`{{html_element("input")}}` sounds unambiguous while `{{js("while")}}` sounds totally weird, so I can symphasize. `{{html_element_xref("input")}}` is fine also but that's unncessarily long. In general I think when we link to a specific part of the reference we can omit `xref`, but when the macro links to everything under the reference we have to disambiguate. Assuming we have different `js_global`, `js_operator`, and `js_statement` macros, the `xref` would not be necessary either. Same for `css_function`, `css_pseudo_class`, etc. (In fact that may have been much better to avoid the ugly `{{cssxref("&lt;length&gt;")}}`.)

> Why are some of our banner/header macros called thingheader and some aren't?

The unfortunate exception is `{{SeeCompatTable}}`, I think? I recall a PR that renamed it to `{{experimental_header}}` but that PR went nowhere. (Edit: https://github.com/mdn/yari/pull/8259)

**Reply by hamishwillee** [reply]
I have every confidence we could manage this disruption: we've proved that in previous projects. I'm with WillB that it is worth doing all the way or not at all. I don't feel the need to be bound by the existing conventions 
- changing `{{SeeCompatTable}}` to `{{experimental_header}}` is a great example of a desirable change
- While we're at it, I agree with @yin1999 above that consistency in the way arguments are presented would be good: https://github.com/orgs/mdn/discussions/448#discussioncomment-6965324.
-  I would prefer we don't have too many macros - because the more we have the more I have to think. I am very supportive though of something that allows us to remove awkward declaration like `{{cssxref("&lt;length&gt;")}}`

It sounds like no one is strongly opposed and some of us are quite keen. Perhaps the next step would be to add some proposals to the spreadsheet for what we'd like to see? Could add extra rows for possible new macros.

**Reply by hamishwillee** [reply]
I've added two columns to [Will's spreadsheet](https://docs.google.com/spreadsheets/d/1oXrwRhLRM745z0QOAm53ezaPcI55SId9gsFV4ffA7gQ/edit#gid=262444114). The first is an empty column for what we agree on. The second is my column of suggestions (incomplete). I suggest if you have a strong feeling that the suggestions are wrong you add a column on your preference.

Even though there are arguments for things being too long or whatever, I would prefer we do use a naming convention - so 
- xref for anything that is an xref -  domxref, cssxref or dom_xref, css_xref, 
- sidebar for anything that is a sidebar: js_sidebar, apiref_sidebar (or api_ref_sidebar).
- list for lists
- inline for inline badges
- header for headers
- I prefer we use proper snake case for separating words by default - even for domxref etc. So secure_context_header. 

At the end of using above as a default we would then have a set of things we could decide "can't live with" or "OK" to discuss.

> **Note** Renaming macros involves the yari team, who are busy people and not at our beck and call. So it might be we start simple - with those that we agree as a team can just be lower cased and used "as is", and any that are just unacceptable - such as "SeeCompatTable".

**Reply by Josh-Cena** [reply]
> xref for anything that is an xref - domxref, cssxref or dom_xref, css_xref,

I would use "xref" if we are linking to the entire reference section. However I would keep `html_element` and `http_method` et al. as-is. Just a slight optimization.

> sidebar for anything that is a sidebar: js_sidebar, apiref_sidebar (or api_ref_sidebar).

Agreed. I wonder if we can remove the "ref" where appropriate.

> any that are just unacceptable - such as "SeeCompatTable".

I wouldn't call that unacceptable—it was mainly because there was belief that the macro would eventually be replaced with automation. I don't think that's happening soon, though, so we can probably do it within a bigger effort.

**Reply by OnkarRuikar** [reply]
> I think? I recall a PR that renamed it to {{experimental_header}} but that PR went nowhere. (Edit: https://github.com/mdn/yari/pull/8259)

We shouldn't wait for the automation. For now we can reopen the PR or create a new one to change all the macros we want to rename at yari side.

> IMO we should use machines to test the change instead of relying on human review.

I think locally we can test it by building all the pages and looking at the flaws.

> getting people used to using new names - is harder to quantify, but I think it would be OK. We will get used to the names quickly enough.

We can execute it like this:
1. We document the old vs new names.
2. Put a macro linting mechanism (with auto fix capability) in place first. Josh is working on the macro linter/checker. We add macros to the linter config as we change them. After a month we stop auto updating the names.
3. Set the linter to run in Husky and Reviewdog CI for auto corrections.
4. Roll out content side only changes first over a weekend. And add those to the linter config.
5. Send PRs to yari and content to rename macros. When yari team is ready merge content PR and immediately merge yari PR. And add the macros to the linter config.

**Reply by hamishwillee** [reply]
I'd be happy to start by lower casing all of the snake cased ones for which it would work - and happy to do that if nothing else gets agreed. This would be good to do even if "in principle" we want a slightly different convention in future that requires "proper" renaming, because we don't know how long it would take to get that achieved.

Which is why I just merged @Josh-Cena changes in JavaScript area - they are no worse, and to my mind most of them are better (caveat being that he has to change them if we agree some other standard here).

>> IMO we should use machines to test the change instead of relying on human review.

> I think locally we can test it by building all the pages and looking at the flaws.

@wbamberg is suggesting that if the changes do not affect rendering you should be able to build the files before and after the change and do a strict compare. If that works you wouldn't have to look at the content at all.

---

Would you all be amenable to a first phase which is just changes that don't require any involvement of the dev team?

So lower casing an agreed set of items that are already snake cased?
- domxref
- glosssary
- deprecated_inline
- deprecated_header
- etc?

**Reply by Josh-Cena** [reply]
> So lower casing an agreed set of items that are already snake cased?

I'm on board for that. I think my PRs were mostly in that direction except for the `Header` ones due to the BCD sync script.

**Reply by OnkarRuikar** [reply]
> I think my PRs were mostly in that direction except for the Header ones due to the BCD sync script.

You go ahead and use lower cases. I'll update the script to use lower case letters before Saturday release. Other renaming changes like `non-standard_header` to `non_standard_header` will be done once yari is updated. We can manually update those PRs before merging if anything goes wrong.

**Reply by hamishwillee** [reply]
Great. Though we should list the ones we agree, and tick them off the list.

**Reply by hamishwillee** [reply]
I guess I should also add that there is no pressing urgency on this. It doesn't have to be done instantly - it just needs to be co-ordinated such that everyone knows what is happening and the cost to the team is minimized. I would certainly be looking at Will's idea of testing the rendered output as that means pretty much zero review time.

**Reply by wbamberg** [reply]
> Perhaps the next step would be to add some proposals to the spreadsheet for what we'd like to see?

Thanks @hamishwillee and @Josh-Cena for adding some suggestions! I love a good fight over naming. The first sheet in https://docs.google.com/spreadsheets/d/1oXrwRhLRM745z0QOAm53ezaPcI55SId9gsFV4ffA7gQ/edit#gid=262444114 lists all the different casing variations under which macros are currently called, so perhaps isn't the best list to use when discussing how to rename them (because, for instance, it's possible to for someone to make different choices for `domxref` and `DOMXref`, which doesn't make sense).

So I added a new sheet to https://docs.google.com/spreadsheets/d/1oXrwRhLRM745z0QOAm53ezaPcI55SId9gsFV4ffA7gQ/edit#gid=262444114, which just lists each macro under its normalized name. Perhaps you could relocate your suggestions to that sheet?

**Reply by wbamberg** [reply]
I saw that there are suggestions in the sheet to change/merge/delete some macros. I'd prefer to limit this exercise to case normalization and renaming, for one thing because it would make automated testing much easier.

**Reply by hamishwillee** [reply]
1. Yes to let's not delete just yet.
2. I think we are broadly agreed where we want to head _eventually_: lower case snake case
   - I am **mostly** in agreement with @Josh-Cena column Will's [new sheet](https://docs.google.com/spreadsheets/d/1oXrwRhLRM745z0QOAm53ezaPcI55SId9gsFV4ffA7gQ/edit#gid=262444114,)
   - But it would be good to agree a naming convention - I like the idea of having xref, sidebar, list, header, inline, in name to tell you what the thing is for. Further, an order for this: jsxref vs js_xref vs xref_js.

I would like to simplify this problem by doing a first phase that makes all macros of the same type consistent. Right now that is a benefit we can add without any changes required in yari of just having one version, even if it isn't quite where we want to end up. I think consistency is worth the churn.

The suggestion is that where we can lower-case a macro and it is clear and useful we settle on that even if it might change later. Where we can't because the macro is camelcased, we select the best option and make them all match that camel case. 

I have added another column "Phase1" which I propose we do now. It is more or less what you'd expect domxref, jsref, deprecated_inline, for the cases that work, and HTMLElement, APIRef for the cases that don't. 

Can we agree to do phase 1 as outlined. I think that will make talking about phase 2 much easier.

**Reply by hamishwillee** [reply]
PS Does anyone want to test Will's theory that we can do a binary compare of before/after changes for generated content to validate macro changes?

If no one else I guess I can - is it just `yari build` to create a local tree?

**Reply by Josh-Cena** [reply]
@hamishwillee For the sheet, could you use the new sheet instead? You can add two columns for the two phases if you wish.

**Reply by hamishwillee** [reply]
@Josh-Cena No. Because my phase 1 suggestion only makes sense in the context of the original list. 

Getting phase 2 done is going to be a lot harder, and might not happen. If we concentrate on getting all the duplicates gone we'll be achieving something good.

**Reply by Josh-Cena** [reply]
Doing it on the original list easily creates inconsistencies, though :( Deduplicated ones are easier to scan.

**Reply by hamishwillee** [reply]
Nope. I grouped all the original same-items together in the original list to avoid this. You can sort on [not change] to find the results/thing to change to at the end of phase one. 

Or if you want I can just copy the final list here. 

But if the goal of phase 1 is to choose what we settle on before talking about yari, it doesn't make sense to move the lists into the second sheet.

**Reply by Josh-Cena** [reply]
Okay, I tried porting your list to the condensed version. There were a few uncertainties and inconsistencies, so we can discuss about those, but overall I think the content-only phase I should be okay once we have consensus. Unfortunately I can't be in editorial meetings for the entire semester so we have to discuss this async at some moment.

**Reply by hamishwillee** [reply]
Thanks The phase one list pretty much changes to the snake case it that works as our planned end point, and otherwise picks the most popular option if snake casing is going to require a yari change. In other words it is the minimal churn option for changes to macros in phase 1.
- Except that if we decide we will eventually be changing to `js_xref` (and friends) it's a double change.
- How do we feel about this set for phase 1.?
- Do we think a phase 1 is a good idea or should we just finish up agreeing phase 2 and do it all in one go?

I'm quite supportive of my phase 1 set :-)

**Reply by hamishwillee** [reply]
Re ^^^ whether we consider this in 1 phase or 2 really depends on whether we can get changes to the macros approved. If this is in doubt 2 phases make sense. If not we should just agree the final point and work towards it.

My assumption is that given this is a name change we can create new macros for the changed name versions, make content changes, then remove the old versions. Anyone can submit a copied macro with a new name so that's doable. You have rights to merge right @wbamberg ? If that approach is doable I'll stop talking phase 1.

**Reply by wbamberg** [reply]
>  You have rights to merge right @wbamberg ? 

Yes.

I'm in favour of doing it all in one go I think, for 2 reasons:

- it will be painful and scary, and one round of that is better than 2. It's like when you go to the dentist and they do all your root canals at once.
- doing things in 2 phases always makes me wonder, what if we don't get to phase 2? And then in this case we will be stuck with case-sensitivity, but half in camel case and half in snake case, which for writers is in some ways worse than case-insensitivity.

However! Although I'm happy to help support this I won't be able to put time into planning it out, I just have too much else going on at the moment. I think y'all are making good choices about it though 👍 .

**Reply by hamishwillee** [reply]
@wbamberg Thanks. I deleted my phase 1 column
- HW column is what I think we might end up with IF we can test changes by building and binary compare.
- HW invert column puts the category as a prefix
- Now we are thinking churn is not an issue then does anyone very strongly have an opinion for/against CamelCase? I don't care as long as we are consistent.
- I added "_link" as an identifier for previous, next, glossary and so on
- proposing _info as category for anything that pulls in data - such as css syntax, inheritance, compat, etc.
- I moved embed_ to be a suffix for consistency
- propose we prefix internal_ to tag things we don't really expect people to use directly

Or to put it another way given that we are now saying we might suck up the churn then our previous decisions need to be revisited:
- CamelCase vs snake case: JsXref vs js_xref (I prefer snake - less compact but all lower case removes decisions)
- Prefix vs suffix for type - i.e. xref_js/XrefJs, vs js_xref
- All use type identifier, or only some - html_element or html_element_xref
- What types do we prefer - OK with link, embed, info, internal, xref, inline, header, sidebar?

**Reply by yin1999** [reply]
The "Total need call-site changing for snake case" field has not included the macro calls in translated-content. I think this would be a huge task.

**Reply by hamishwillee** [reply]
@yin1999 if we do it then largely an automated process both in terms of conversion and testing. So perhaps not as scary as you think

**Comment by OnkarRuikar** [comment]
Regarding other macro related suff.
1. I prefer double quotes over single quotes:
```diff
-- {{Macro('arg')}}
++ {{Macro("arg")}}
```

2. Explicit double quotes:
```diff
-- {{EmbedLiveSample("audio", 250, 250)}}
++ {{EmbedLiveSample("audio", "250", "250")}}
```

3. No brackets for zero arguments:
```diff
-- {{CSSref()}}
++ {{CSSref}}
```

4. No space between macro call and curly braces:
```diff
-- {{ APIRef }} {{ Non-standard_header }} {{ deprecated_header }}
++ {{APIRef}} {{Non-standard_header}} {{deprecated_header}}
```

5. Explicit use of units:
```diff
-- {{EmbedLiveSample("audio", "100", "70")}}
++ {{EmbedLiveSample("audio", "100%", "70px")}}
```

No preference for following cases:
6. Empty params:
```diff
-- {{EmbedLiveSample('audio', '', '70px')}}
++ {{EmbedLiveSample('audio', '100%', '70px')}}
```

7. Sentence or single word:
```diff
-- {{EmbedLiveSample("Audio example")}}
++ {{EmbedLiveSample("audio_example")}}
```

**Reply by hamishwillee** [reply]
- Agree 1-4 inclusive. 
- 5, 6: I prefer units but happy to have on hold it it requires a change to Yari? Yes, empty params need to be marked.
- 7: I prefer sentence for the "7" because then I just copy in the heading text.

**Reply by OnkarRuikar** [reply]
> happy to have on hold it it requires a change to Yari?

Yari sets the width and height arguments [as it is on iframe](https://github.com/mdn/yari/blob/76f8d444e4e112ce49504e09ec9376445a10953e/kumascript/macros/EmbedLiveSample.ejs#L91-L92). Treat it as `<iframe height="${arg}">` So no change on Yari side is needed.

**Comment by OnkarRuikar** [comment]
> PS Does anyone want to test Will's theory that we can do a binary compare of before/after changes for generated content to validate macro changes?
>
> If no one else I guess I can - is it just yari build to create a local tree?

Using command line we can use following shell script:
```bash
# 1. get fresh `mdn/content` code

# 2. assuming you've the latest yari code and done `yarn install`
cd /projects/mdn/yari
yarn build --locale en-us
mv client/build /temp_dir/build

# 3. do the macro renaming in `mdn/content`

# 4. build content again
yarn build --locale en-us

# 5. list the files that are different
diff -qr /temp_dir/build/ client/build > different_files.log
```
On Windows I think this should run without any issues in [GIT BASH](https://gitforwindows.org/).
For step 5 you can use GUI diff tool [Meld](https://meldmerge.org/#features) for [dir comparison](https://meldmerge.org/images/dirdiff.png).

**Reply by hamishwillee** [reply]
FYI using normal build (i.e. not Yari dev) two builds are identical for most normal content after macro changes in case. So the approach appears to work (Windows).

**Reply by OnkarRuikar** [reply]
I am using this to test "CSS Formal syntax" code. It is very useful to see all the corner cases. 
I am using following version to see exact differences:
```bash
# 1. get fresh `mdn/content` code

# 2. assuming you've the latest yari code and done `yarn install`
cd /projects/mdn/yari
yarn build --locale en-us
cd client/build
rm -r **/**/metadata.json
rm -r **/**/index.json
// with above deletes it takes 19min for Prettier. Otherwise it takes 27min.
prettier --write "**/**/index.html"
cd ../..
mv client/build/en-US /temp_dir/old_en-US

# 3. update the CSS formal syntax code

# 4. build content again
yarn build --locale en-us
cd client/build
rm -r **/**/metadata.json
rm -r **/**/index.json
prettier --write "**/**/index.html"
cd ../..

# 5. Use Meld tool to compare folders and see exact changes.
# 6. Do search replace on all built files if there are any new hashes in script file names or any other same changes in all files.
```

**Reply by hamishwillee** [reply]
I'm too lazy. Result is that the approach will work or not?

**Reply by OnkarRuikar** [reply]
> I'm too lazy. Result is that the approach will work or not?

It is working for me in Linux setup. If you want give me the changed content and I'll run the test.

**Reply by hamishwillee** [reply]
Cool. We'll have to agree the final set and then create some macros etc before we can get that far. Someone will also have to write a conversion script.

**Reply by OnkarRuikar** [reply]
> Cool. We'll have to agree the final set 

Do we need a meeting for this?

> Someone will also have to write a conversion script.

Simple regex search replace in any IDE should do it.

**Reply by hamishwillee** [reply]
@OnkarRuikar I was thinking maybe a bit more than basic regexp for the bits you identified here https://github.com/orgs/mdn/discussions/448#discussioncomment-6963747

I think the next step is that I will create a column "Agreed changes" and see if anyone disagrees with any of them :-)

But probably not until Monday

**Comment by hamishwillee** [comment]
So in the sheet there were are few contentions around the xref and suffixes
- jsxref (JC) or js_xref (WB)
- glossary (JC) or glossary_xref.
- html_element (JC), html_element_xref (HW)

1. A suffix (or prefix) is supposed to differentiate the type. You might not think `html_element_xref` is needed, but a newbie could think that `html_element` means a sidebar for html elements, or a list of html element.
  So IMO we either have suffixes and use them consistently or not at all. 
2. Could we have no xref suffix? So `js`, `css`, `api`, etc. Maybe, if we consistently have an extension for every other type `_sidebar`, `_list` and so on. I personally think it is better to have it and be consistent.
3. I prefer `js_xref` to `jsxref` for consistency reasons - same pattern for all suffixes.

**Reply by OnkarRuikar** [reply]
> Could we have no xref suffix? So `js`, `css`, `api`, etc. Maybe, if we consistently have an extension for every other type _sidebar, _list and so on. I personally think it is better to have it and be consistent.

+1 If everything is defined then we can have these shorter versions.\
Otherwise `js_xref`, `glossry_xref` sounds good to me. Can we use `html_xref` instead of `html_element_xref`?

@hamishwillee Regarding your comment in the sheet, `CSS_ref` is a valid macro and has been used in https://developer.mozilla.org/en-US/docs/Web/CSS/reference#index . It can be renamed to `CSS_index`, because at the moment there are `CSSref` and `CSS_ref` which is confusing.

Also for clarity following could be renamed:
ListGroups -> api_groups_list
APIListAlpha -> api_interfaces_list
ListSubpages -> sub_pages_list
ListSubpagesForSidebar -> sub_pages_list_for_sidebar

**Reply by Josh-Cena** [reply]
`html_xref` doesn't make sense because you are only linking to elements, not attributes. I'm all for `htmlxref` if we end up having the same macro that replaces both `html_element` and `html_attr`.

`glossry_xref` also doesn't make sense because `xref` means "link to reference", but glossary is not a reference.

The idea is this: if we are linking to a "reference section" as a whole, and the section contains many parallel sets of topics, then we add the `xref` suffix. For example:

- The entire web API reference is one flat list of the API interfaces. Therefore `api("fetch")` by itself is good. One complication is that there are also guide pages mixed among these, so we either change the content structure or prevent them from being linkable to keep the "purity" of the `api` macro.
- The JS reference contains disjoint sets of documents for globals, statements, and expressions. Therefore, the name `js` by itself tells you nothing about the semantics of the page being linked to (what is `js("class")`?) In this case, we add the suffix `xref` to be explicit that it's a link to the entire reference. We may also use `js_global("Array")` and `js_statement("while")`, but I'm sure people won't want that.
- The CSS reference also contains disjoint sets of documents for pseudoclasses, functions, selectors, etc. Therefore it's the same situation and we add `xref`.
- The glossary is (a) not a reference (b) contains a flat list of term definitions, therefore `glossary("identifier")` is good enough.
- The HTML reference contains documents for elements and attributes. It happens that there are also two macros—`html_element` and `html_attr`—bravo!

I hope that makes sense?

**Reply by Josh-Cena** [reply]
I don't feel too strongly about `jsxref` vs `js_xref`, as I also commented in the doc. My main arguments are:

- Significantly less churn (xrefs are by far the most popular macros and are ever growing in number)
- Easier authoring experience (takes much less time to type a letter compared to an underscore)

But I can relate: I used to think it means `jsx_ref`, because JSX is an actual thing. However, that doesn't really affect how I use the macro; just a bit of misunderstanding.

**Reply by hamishwillee** [reply]
I'm not so concerned about the churn if we are able to get this checked using an approach like @OnkarRuikar tried here https://github.com/orgs/mdn/discussions/448#discussioncomment-7009651.

**Reply by hamishwillee** [reply]
PS I tried that approach and it seems to work. @Josh-Cena I had some more questions above https://github.com/orgs/mdn/discussions/448#discussioncomment-7097778 - basically we're all still happy with suffixes rather than prefixes? I think we're happy with js_xref vs jsxref too.

**Reply by Josh-Cena** [reply]
I don't feel strongly about using prefix or suffix, although I don't see a strong motivation to change the existing ones either. `api_sidebar` does read more naturally than `sidebar_api` to me. However I feel like `compat` => `compat_info` doesn't add much meaning. Usually the suffix "info" is not needed; it's only added for `css_info` because `css` doesn't make sense. If we decided to use prefix, then `xref_js` would be good, but in the status quo I want to keep `jsxref`,
